File: question3.py
# delete node in a linked list

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def append(head, data):
    if head is None:
        return Node(data)
    temp = head
    while temp.next:
        temp = temp.next
    temp.next = Node(data)
    return head

def deleteNode(head, x):
    if head is None:
        return head

    if x == 1:
        return head.next

    temp = head
    for _ in range(x - 2):
        if temp is None or temp.next is None:
            return head
        temp = temp.next

    if temp.next is None:
        return head

    temp.next = temp.next.next
    return head

def print_list(head):
    temp = head
    while temp:
        print(temp.data, end=" ")
        temp = temp.next
    print()


n = int(input("Enter number of nodes: "))
head = None

print("Enter the elements:")
for _ in range(n):
    val = int(input())
    head = append(head, val)

x = int(input("Enter the position to delete (1-based index): "))

head = deleteNode(head, x)

print("Linked list after deletion:")
print_list(head)

File: question3.c
// delete node in a linked list

#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

void appendNode(struct Node** head, int data) {
    struct Node* newNode = createNode(data);
    if (*head == NULL) {
        *head = newNode;
        return;
    }
    struct Node* temp = *head;
    while (temp->next) {
        temp = temp->next;
    }
    temp->next = newNode;
}


struct Node* deleteNode(struct Node* head, int x) {
    if (head == NULL) return head;

    if (x == 1) {
        struct Node* temp = head;
        head = head->next;
        free(temp);
        return head;
    }

    struct Node* temp = head;
    for (int i = 1; temp != NULL && i < x - 1; i++) {
        temp = temp->next;
    }

    if (temp == NULL || temp->next == NULL) return head;

    struct Node* nodeToDelete = temp->next;
    temp->next = nodeToDelete->next;
    free(nodeToDelete);

    return head;
}

void printList(struct Node* head) {
    while (head != NULL) {
        printf("%d ", head->data);
        head = head->next;
    }
    printf("\n");
}

int main() {
    struct Node* head = NULL;
    int n, val, x;

    printf("Enter number of nodes: ");
    scanf("%d", &n);

    printf("Enter the elements:\n");
    for (int i = 0; i < n; i++) {
        scanf("%d", &val);
        appendNode(&head, val);
    }

    printf("Enter the position to delete (1-based index): ");
    scanf("%d", &x);

    head = deleteNode(head, x);

    printf("Linked list after deletion:\n");
    printList(head);

}
